crash before reaching my main

i'm trying to compile a simple imgui app with cargo-apk. after successfully working through some linking issues(__dso_handle etc.), and manually setting uses_cpp_standard_library to true, i finally got the apk to run but it crashes immediately.

> -04 19:11:56.776  9852  9865 D RustAndroidGlueStdouterr: Entering android_main
> 09-04 19:11:56.776  9852  9865 D RustAndroidGlueStdouterr: Creating application thread
> 09-04 19:11:56.778  9852  9866 D RustAndroidGlueStdouterr: thread '<unnamed>' panicked at 'assertion failed: (*self.data.get()).is_none()', src/libstd/sync/mpsc/oneshot.rs:90:13

before reaching my main. so i could not set RUST_BACKTRACE.....

i set android_version, target_sdk_version and min_sdk_version to 24.

i use the version from git, it uses android-glue 0.2.3. i'm using android 7.0, perhaps this is the issue.

it now crashes on the vulkan side, so this issue seems to be solved for me, after using android-glue current ;)

i tried to reproduce my linking errors(undefined reference to __dso_handle, can not not reposition ... use -FPIC), but after cleaning everything, they have gone away.

as for the uses_cpp_standard_library issue, i have honestly no understanding, how cargo-apk should even know when to include libc++_shared. 

when i include a rustflags = ["-C", "link-arg=-lc++"] in .cargo/config, the parameters get passed to rustc, but i still have no libstdc++_shared included.

i tried to add a .cpp_link_stdlib("c++") to the build.rs of the imgui-sys crate - still does not work.

so i simply set uses_cpp_standard_library to true, which does work.

--- a/cargo-apk/src/ops/build/compile.rs
+++ b/cargo-apk/src/ops/build/compile.rs
@@ -7,7 +7,7 @@ use cargo::core::compiler::Executor;
 use cargo::core::manifest::TargetSourcePath;
 use cargo::core::{PackageId, Target, TargetKind, Workspace};
 use cargo::util::command_prelude::ArgMatchesExt;
-use cargo::util::{CargoResult, ProcessBuilder};
+use cargo::util::{process, CargoResult, ProcessBuilder};
 use clap::ArgMatches;
 use failure::format_err;
 use multimap::MultiMap;
@@ -318,24 +318,25 @@ pub extern "C" fn android_main(app: *mut ()) {{
                 target.clone(),
                 SharedLibrary {
                     abi: self.build_target,
-                    path: library_path,
+                    path: library_path.clone(),
                     filename: format!("lib{}.so", target.name()),
                 },
             );
 
             // If the target uses the C++ standard library, add the appropriate shared library
             // to the list of shared libraries to be added to the APK
-            let mut iter = new_args.iter().peekable();
-            let mut uses_cpp_standard_library = false;
-            while let Some(arg) = iter.next() {
-                if let Some(next_arg) = iter.peek() {
-                    if arg == "-l" && *next_arg == "c++" {
-                        uses_cpp_standard_library = true;
-                    }
-                }
-            }
+            let readelf_path = util::find_readelf(&self.config, self.build_target)?;
+            let readelf_output = process(readelf_path)
+                .arg("-d")
+                .arg(&library_path)
+                .exec_with_output()?;
+            use std::io::BufRead;
+            let dynamically_links_to_cpp_standard_lib = readelf_output.stdout.lines().any(|l| {
+                let l = l.as_ref().unwrap();
+                l.contains("(NEEDED)") && l.contains("[libc++_shared.so]")
+            });
 
-            if uses_cpp_standard_library {
+            if dynamically_links_to_cpp_standard_lib {
                 let cpp_library_path = version_independent_libraries_path.join("libc++_shared.so");
                 shared_libraries.insert(
                     target.clone(),

--- a/cargo-apk/src/ops/build/util.rs
+++ b/cargo-apk/src/ops/build/util.rs
@@ -120,7 +120,7 @@ pub fn find_clang_cpp(
     .map_err(|_| format_err!("Unable to find NDK clang++"))
 }
 
-// Returns path toe ar tool.
+// Returns path to ar.
 pub fn find_ar(config: &AndroidConfig, build_target: AndroidBuildTarget) -> CargoResult<PathBuf> {
     let ar_path = llvm_toolchain_root(config).join("bin").join(format!(
         "{}-ar{}",
@@ -131,8 +131,28 @@ pub fn find_ar(config: &AndroidConfig, build_target: AndroidBuildTarget) -> Carg
         Ok(ar_path)
     } else {
         Err(format_err!(
-            "Unable to find AR file at `{}`",
+            "Unable to find ar at `{}`",
             ar_path.to_string_lossy()
+        ))
+    }
+}
+
+// Returns path to readelf
+pub fn find_readelf(
+    config: &AndroidConfig,
+    build_target: AndroidBuildTarget,
+) -> CargoResult<PathBuf> {
+    let readelf_path = llvm_toolchain_root(config).join("bin").join(format!(
+        "{}-readelf{}",
+        build_target.ndk_triple(),
+        EXECUTABLE_SUFFIX_EXE
+    ));
+    if readelf_path.exists() {
+        Ok(readelf_path)
+    } else {
+        Err(format_err!(
+            "Unable to find readelf at `{}`",
+            readelf_path.to_string_lossy()
         ))
     }
 }

